### PR TITLE
TRestPhysics.h added parabolic/hyperbolic interaction

### PR DESCRIPTION
**Pull request by jovoy**
jovoy Medium: 100 [](https://gitlab.cern.ch/rest-for-physics/framework/-/commits/true_wolter_interaction) [](https://github.com/rest-for-physics/framework/commits/true_wolter_interaction)

Adding the interaction between a vector and a parabolic/hyperbolic shaped mirror:

- [x] Find interaction point between vector and parabolic mirror
- [x] Find interaction point between vector and hyperbolic mirror
- [x] normal vector on parabolic mirror
- [x] normal vector on hyperbolic mirror
- [x] the latter need to be normalized for further calculations

### CODE REVIEW

**Comment by jgalan** [comment]
Hi @jovoy I just pushed a commit to try to fix the compilation issues, and fix the pipeline.

File: source/framework/tools/src/TRestPhysics.cxx [hunk 1]
```diff
@@ -139,6 +195,41 @@ TVector3 GetConeNormal(const TVector3& pos, const Double_t& alpha, const Double_
...
+/// `beta` is the angle between the hyperboloid and the optical (z) axis at the plane where the
+/// hyperboloid has the radius `R3`.
+///
+TVector3 GetHyperbolicNormal(const TVector3& pos, const Double_t& alpha, const Double_t& R3,
+                             const Double_t& focal) {
+    Double_t z = 1;  /// This is just a temp value to fix compilation issues
+    TVector3 normalVec = pos;
+    Double_t beta = 3 * alpha;
+    /// Just replaced here *TMath::Cot by /TMath::Tan to fix compilation issues
+    Double_t m = 1 / (R3 * TMath::Tan(beta) * (1 + 2 * (-z) / (focal + R3 / TMath::Tan(2 * alpha))) /
```
Comment (jgalan) [inline source/framework/tools/src/TRestPhysics.cxx:225]:
```suggestion
    Double_t m = 1 / (R3 * TMath::Tan(beta) * (1 - 2 * pos.Z() / (focal + R3 / TMath::Tan(2 * alpha))) /
```

File: source/framework/tools/src/TRestPhysics.cxx [hunk 2]
```diff
@@ -139,6 +195,41 @@ TVector3 GetConeNormal(const TVector3& pos, const Double_t& alpha, const Double_
...
+
+///////////////////////////////////////////////
+/// \brief This method returns the normal vector on a hyperbolic surface pointing towards the inside
+/// of the hyperboloid. `pos` is the origin point of the normal vector on the hyperbolic plane and
+/// `beta` is the angle between the hyperboloid and the optical (z) axis at the plane where the
+/// hyperboloid has the radius `R3`.
+///
+TVector3 GetHyperbolicNormal(const TVector3& pos, const Double_t& alpha, const Double_t& R3,
+                             const Double_t& focal) {
+    Double_t z = 1;  /// This is just a temp value to fix compilation issues
```
Comment (jgalan) [inline source/framework/tools/src/TRestPhysics.cxx:221]:
```suggestion
```

File: source/framework/tools/src/TRestPhysics.cxx [hunk 3]
```diff
@@ -139,6 +195,41 @@ TVector3 GetConeNormal(const TVector3& pos, const Double_t& alpha, const Double_
...
+    TVector3 normalVec = pos;
+    Double_t beta = 3 * alpha;
+    /// Just replaced here *TMath::Cot by /TMath::Tan to fix compilation issues
+    Double_t m = 1 / (R3 * TMath::Tan(beta) * (1 + 2 * (-z) / (focal + R3 / TMath::Tan(2 * alpha))) /
+                      TMath::Sqrt(R3 * R3 + R3 * 2 * TMath::Tan(beta) * (-z) *
```
Comment (jgalan) [inline source/framework/tools/src/TRestPhysics.cxx:226]:
```suggestion
                      TMath::Sqrt(R3 * R3 - R3 * 2 * TMath::Tan(beta) * pos.Z() *
```

File: source/framework/tools/src/TRestPhysics.cxx [hunk 4]
```diff
@@ -139,6 +195,41 @@ TVector3 GetConeNormal(const TVector3& pos, const Double_t& alpha, const Double_
...
+                                                (1.0 + (-z) / (focal + R3 / TMath::Tan(2 * alpha)))));
```
Comment (jgalan) [inline source/framework/tools/src/TRestPhysics.cxx:227]:
```suggestion
                                                (1.0 - pos.Z() / (focal + R3 / TMath::Tan(2 * alpha)))));
```

**Review by jgalan — APPROVED** [review]

File: source/framework/tools/inc/TRestPhysics.h [hunk 5]
```diff
@@ -76,8 +76,18 @@ Double_t GetVectorsAngle(const TVector3& v1, const TVector3& v2);
 
 TVector3 GetPlaneVectorIntersection(const TVector3& pos, const TVector3& dir, TVector3 const& n,
                                     TVector3 const& a);
+  
+TVector3 GetParabolicVectorIntersection(const TVector3& pos, const TVector3& dir, const Double_t& alpha,
+                                         const Double_t& R3, const Double_t& lMirr);
```
Comment (Vindaar) [inline source/framework/tools/inc/TRestPhysics.h:81]:
```suggestion
                                        const Double_t& R3, const Double_t& lMirr);
```

File: source/framework/tools/src/TRestPhysics.cxx [hunk 6]
```diff
@@ -83,6 +83,62 @@ TVector3 GetPlaneVectorIntersection(const TVector3& pos, const TVector3& dir, co
     return MoveToPlane(pos, dir, n, a);
 }
 
+//////////////////////////////////////////////
+/// This method will find the intersection between a vector and a parabolic shape where `alpha` is the angle
+/// between the optical axis and the paraboloid at the plane where the paraboloid has a radius of `R3`.
+/// The paraboloid is rotational symmetric around the optical axis. `Alpha` in rad.
```
Comment (Vindaar) [inline source/framework/tools/src/TRestPhysics.cxx:89]:
```suggestion
/// The paraboloid is rotationally symmetric around the optical axis. `alpha` in rad.
```

File: source/framework/tools/src/TRestPhysics.cxx [hunk 7]
```diff
@@ -83,6 +83,62 @@ TVector3 GetPlaneVectorIntersection(const TVector3& pos, const TVector3& dir, co
...
+    } else if (pos.Z() + root2 * dir.Z() > -lMirr and pos.Z() + root2 * dir.Z() < 0) {
+        return pos + root2 * dir;
+    }
+    return pos;
+}
+
+//////////////////////////////////////////////
+/// This method will find the intersection between a vector and a hyperbolic shape where 3 * `alpha` is the
+/// angle between the optical axis and the hyperboloid at the plane where the hyperboloid has a radius of
+/// `R3`. The hyperboloid is rotational symmetric around the optical axis. `Alpha` in rad. hTe region in which
```
Comment (Vindaar) [inline source/framework/tools/src/TRestPhysics.cxx:115]:
```suggestion
/// `R3`. The hyperboloid is rotationally symmetric around the optical axis. `alpha` in rad. The region in which
```

**Review by Vindaar — APPROVED** [review]
Just a few minor typos. Otherwise looks good (even if `root1` and `root2` could ideally use a separately computed value for the `sqrt`, but that's being extra pedantic of course :) ).

**Pull request merged**